### DISCUSSION TITLE
code/style reuse

### DISCUSSION THREAD

**Discussion by schmurfy** [terrastruct/d2#1087] [General]
d2 is nice but for complex diagram you need to repeat everything, is there any plan for an easier way to reuse shapes, styles, ... ?
For example I would like to to something like:
```
#service {
  shape: cloud
  style {
    fill: lightblue
    stroke: green
  }
}

service1 #service
service2 #service
service3 #service
```

**Comment by alixander** [comment]
@schmurfy yeahh we'll have classes. i've been meaning to work on this for a while now. will try to get it in this coming week.

what the syntax will look like:
https://github.com/terrastruct/d2/pull/772/files

**Reply by schmurfy** [reply]
nice, glad to know it is in the pipelines, d2 is really a great tool :)

**Reply by kamelohr** [reply]
@alixander Hi, I'm working in the design field and if you could manage to introduce that feature it would be a game changer for my team and me. ATM it's not really feasible as we'd have to re-define styles on every element.

Additionally it would be awesome if we could define default styles that are then used in any element inside a document as long as the specific style attribute is not overridden through a style class or inside the element itself. Something similar to CSS's 

`* {
/* Some attributes */
}
`

**Reply by alixander** [reply]
this week for sureeee, sorry for the delays everyone and thank you for bumping @kamelohr 

(related github issue: https://github.com/terrastruct/d2/issues/102)

**Reply by alixander** [reply]
> Additionally it would be awesome if we could define default styles that are then used in any element

my colleague @nhooyr is working on globs which will be the mechanism to do that

**Comment by alixander** [comment]
classes will be in next release: https://github.com/terrastruct/d2/pull/772

**Comment by alixander** [comment]
classes is now live in 0.4.0!
